main/cli.c

```diff
@@ -1948,6 +1949,64 @@ static char *group_show_channels(struct ast_cli_entry *e, int cmd, struct ast_cl
 #undef FORMAT_STRING
 }
 
+static char *group_show_variables(struct ast_cli_entry *e, int cmd, struct ast_cli_args *a)
+{
+#define FORMAT_STRING     "%-20s  %-20s\n"
+#define FORMAT_STRING_VAR "     %s=%s\n"
+
+	struct ast_group_meta *gmi = NULL;
+	struct varshead *headp;
+	struct ast_var_t *variable = NULL;
+	int numgroups = 0;
+	char group[80];
+	char category[80];
+
+	switch (cmd) {
+	case CLI_INIT:
+		e->command = "group show variables";
+		e->usage =
+			"Usage: group show variables [group@[category]]\n"
+			"       Lists all currently active groups and their variables.\n"
+			"       Optional group, or group@category can be used to only show a specific group\n";
+		return NULL;
+	case CLI_GENERATE:
+		return NULL;
+	}
+
+	if (a->argc < 3 || a->argc > 4) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (a->argc == 4) {
+		if (ast_app_group_split_group(a->argv[3], group, sizeof(group), category, sizeof(category))) {
+			return NULL;
+		}
+	}
+
+	ast_cli(a->fd, "Group\n    Category\nVariables\n");
+	ast_cli(a->fd, "------------------------------\n");
+
+	/* Print group variables */
+	ast_app_group_meta_rdlock();
+	gmi = ast_app_group_meta_head();
+	while (gmi) {
```

Comment:
I think the suggested move would violate historic layout.  No other component implements CLI functions in app.c and no other component implements any kind of command in app.c, (no cli, no dialplan applications, no dialplan functions)

Reply:
I think the suggested move would violate historic layout. No other component implements CLI functions in app.c and no other component implements any kind of command in app.c, (no cli, no dialplan applications, no dialplan functions)

Reply:
Okay, fair enough - I don't really like the ast_app_group_meta functions existing - is there another way we could get rid of them? If not, I guess it's fine.

Reply:
There's already public ast_app_group_list_wrlock/ast_app_group_list_rdlock that's existed for many years.  But also, just because it's there, doesn't mean it's a good design.

Another approach would be implement an iterator, but then you still want an obvious way to grab the lock... I don't thing the locking should be 'hidden' / embedded in the api, because just like channels, you still have to explicitly grab a channel lock and then do your thing to the channel.  And then implementing all this iterator stuff  sounds like re-inventing ao2 to me.

So speaking of ao2... this would require a bit of an overhaul to convert to ao2.   If anything it would add overhead without much in terms of functional gain considering groups and group variables are fairly simple.